man/webp-compress.1
.TH "webp-compress" 1 2.6.4 "14 Feb 2023" "User manual"

.SH NAME
webp-compress

.SH DESCRIPTION
Compress JPEGs by re-encoding to the smallest WEBP quality while keeping perceived
visual quality the same.
This is a lossy operation, but the images are visually identical and it usually
saves >50% of the size for JPEGs coming from a digital camera, particularly DSLRs.
All EXIF/IPTC/XMP and color profile metadata are not preserved!

.SH SYNOPSIS
webp-compress [options] image.[jpg|ppm] compressed.webp

.SH OPTIONS
.TP
\fB\-c\fR, \fB\-\-no-copy\fR
disable copying files that will not be compressed
.TP
\fB\-d\fR, \fB\-\-defish\fR [arg]
set defish strength [0.0]
.TP
\fB\-f\fR, \fB\-\-force\fR
force process
.TP
\fB\-h\fR, \fB\-\-help\fR
output program help
.TP
\fB\-l\fR, \fB\-\-loops\fR [arg]
set the number of runs to attempt [8]
.TP
\fB\-m\fR, \fB\-\-method\fR [arg]
set comparison method to one of 'mpe', 'psnr', 'mse', 'msef', 'cor', 'ssim', 'ms-ssim', 'vifp1', 'smallfry', 'shbad', 'nhw', 'ssimfry', 'ssimshb', 'sum' [sum]
.TP
\fB\-n\fR, \fB\-\-min\fR [arg]
minimum quality [1]
.TP
\fB\-q\fR, \fB\-\-quality\fR [arg]
set a quality preset: low, medium, high, veryhigh [medium]
.TP
\fB\-r\fR, \fB\-\-ppm\fR
parse input as PPM
.TP
\fB\-t\fR, \fB\-\-target\fR [arg]
set target quality UM [0.76]
.TP
\fB\-x\fR, \fB\-\-max\fR [arg]
maximum quality [99]
.TP
\fB\-z\fR, \fB\-\-zoom\fR [arg]
set defish zoom [1.0]
.TP
\fB\-Q\fR, \fB\-\-quiet\fR
only print out errors
.TP
\fB\-T\fR, \fB\-\-input-filetype\fR [arg]
set input file type to one of 'auto', 'jpeg', 'ppm' [auto]
.TP
\fB\-V\fR, \fB\-\-version\fR
output program version

.SH EXAMPLES
Default settings:
.PP
.I
webp-compress image.jpg compressed.webp
.PP
High quality example settings:
.PP
.I
webp-compress --quality high --min 60 image.jpg compressed.webp
.PP
Remove fisheye distortion (Tokina 10-17mm on APS-C @ 10mm):
.PP
.I
webp-compress --defish 2.6 --zoom 1.2 image.jpg defished.webp
.PP
Read from stdin and write to stdout with '-' as the filename:
.PP
.I
webp-compress - - <image.jpg >compressed.webp
.PP
Convert RAW to JPEG via PPM from stdin:
.PP
.I
dcraw -w -q 3 -c IMG_1234.CR2 | webp-compress --ppm - compressed.webp
.PP
Disable all output except for errors:
.PP
.I
webp-compress --quiet image.jpg compressed.webp

.SH NOTES
"Universal Scale" of metrics (UM):
  0.0
  ... (DIRTY) ...
  0.5
  ... (LOW) ...
  0.75
  ... (MEDIUM) ...
  0.875
  ... (SUBHIGH) ...
  0.9375
  ... (HIGH) ...
  0.96875
  ... (VERYHIGH) ...
  1.0
.PP
Trends:

  UM = 0.29 * (sqrt(sqrt(255.0 / MPE)) - 1.0)
  UM = 0.557 * (sqrt(PNSR) - 5.0)
  UM = 0.5 * (sqrt(sqrt(1.0 / MSEF)) - 1.0)
  UM = 1.0 * cor_sigma(cor_sigma(COR))
  UM = 1.57 * cor_sigma(cor_sigma(cor_sigma(SSIM)))
  UM = 1.59 * cor_sigma(cor_sigma(MS_SSIM))
  UM = 1.10 * cor_sigma(cor_sigma(VIFP1))
  UM = 3.0 * (SMALLFRY * 0.01 - 0.8)
  UM = 1.46 * cor_sigma(SHARPENBAD)
  UM = 0.342 * (sqrt(sqrt(1.0 / NHW)) - 1.0)
   cor_sigma(M) = 1.0 - sqrt(1.0 - M * M)

.SH COPYRIGHT
 JPEG-Archive is copyright © 2015 Daniel G. Taylor
 Image Quality Assessment (IQA) is copyright 2011, Tom Distler (http://tdistler.com)
 SmallFry is copyright 2014, Derek Buitenhuis (https://github.com/dwbuiten)
 All rights reserved.

.SH "SEE ALSO"
 jpeg-recompress,
 jpeg-compare,
 jpeg-hash,
 jpeg-zfpoint,
 cjpeg


CHANGELOG
JPEG Recompress
https://github.com/ImageProcessing-ElectronicPublications/jpeg-recompress

2.6.4 "um"

    Research: correct UM model: all metrics.
    2023, zvezdochiot

2.6.3 "um"

    Research: correct UM model.
    2023, zvezdochiot

2.6.2 "sharpenbad"

    reimplementation SHARPENBAD:
      metric_sharpenbad

2.6.1 "msef"

    Added MSE and MSEF metric.
    Research: correct trends UM (44 images)

2.6.0 "sum"

    Added VIFP 1 layer in SUM metric.

2.5.9 "vifp1"

    Added VIFP 1 layer metric.

2.5.8 "neat"

    Neat adds up all values.

2.5.7 "bisection"

    Fix bisection

2.5.6 "subsample"

    Fix IPEP/#1: ignores subsampling option

2.5.5 "sigma"

    Move `sigma`-function from `libsmallfry` to `jmetrics.a`
    Research: correct trends UM (32 images)
    Remove metrics `cor` and `corsh`
    2022, zvezdochiot

2.5.4 "near"

    Research: correct trends UM (24 images)
    2022, zvezdochiot

2.5.3 "approx"

    Research: correct trends UM (16 images)
    2022, zvezdochiot

2.5.2 "UM"

    Research: correct trends UM (8 images)
    2022, zvezdochiot

2.5.1 "saved"

    Fix overflow saved.
    Correct quality scale (next = (1.0 + prev) / 2)
    Replacing multiplications with counters.
    2022, zvezdochiot

2.5.0 "webp"

    Use libwebp for WEBP format
    2022, zvezdochiot

2.4.0 "nhw"

    Update for libsmallfry 0.1.0
    metric_nhw
    2022, zvezdochiot

2.3.1 "weight"

    weight average for SUM metrics
    2021, zvezdochiot

2.3.0 "return"

    return from mozjpeg to jpeg8
    2021, zvezdochiot

2.2.1 "linearize"

    jpeg-recompress: linearize metric
    2018, zvezdochiot

2.2 "zfpoint"

    Find ZF point quality for JPEG
    Update for libsmallfry 0.0.3
    2018, zvezdochiot

2.1.3 "sum"

    jpeg-recompress: summary metric and linearize ssim
    2018, zvezdochiot

2.1.2

    Update for libsmallfry 0.0.2
    sharpenbad_metric
    2018, zvezdochiot

2.1.1

    fork jpeg-archive 2.1.1 to jpeg-recompress for compatible libsmallfry.
    Daniel G. Taylor & zvezdochiot


src/webp-compress.c
/*
    Recompress a JPEG file while attempting to keep visual quality the same
    by using structural similarity (SSIM) as a metric. Does a binary search
    between JPEG quality 40 and 95 to find the best match. Also makes sure
    that huffman tables are optimized if they weren't already.
*/

#include <webp/encode.h>
#include <webp/decode.h>
#include <getopt.h>
#include "jmetrics.h"

const char *COMMENT = "Compressed by webp-compress";

void usage(char *progname)
{
    printf("usage: %s [options] input.[jpg|ppm] output.webp\n\n", progname);
    printf("options:\n\n");
    printf("  -c, --no-copy                disable copying files that will not be compressed\n");
    printf("  -d, --defish [arg]           set defish strength [0.0]\n");
    printf("  -f, --force                  force process\n");
    printf("  -h, --help                   output program help\n");
    printf("  -l, --loops [arg]            set the number of runs to attempt [8]\n");
    printf("  -m, --method [arg]           set comparison method to one of:\n");
    printf("                               'mpe', 'psnr', 'mse', 'msef', 'cor', 'ssim', 'ms-ssim', 'vifp1',\n");
    printf("                               'smallfry', 'shbad', 'nhw', 'ssimfry', 'ssimshb', 'sum' [sum]\n");
    printf("  -n, --min [arg]              minimum quality [1]\n");
    printf("  -q, --quality [arg]          set a quality preset: low, medium, subhigh, high, veryhigh [medium]\n");
    printf("  -r, --ppm                    parse input as PPM\n");
    printf("  -t, --target [arg]           set target quality [0.75]\n");
    printf("  -x, --max [arg]              maximum quality [99]\n");
    printf("  -z, --zoom [arg]             set defish zoom [1.0]\n");
    printf("  -Q, --quiet                  only print out errors\n");
    printf("  -T, --input-filetype [arg]   set input file type to one of 'auto', 'jpeg', 'ppm' [auto]\n");
    printf("  -V, --version                output program version\n");
}

int main (int argc, char **argv)
{
    int method = SUMMET;

    // Number of binary search steps
    int attempts = 8;

    float target = 0.0f;
    int preset = MEDIUM;

    // Min/max JPEG quality
    int qMin = 1;
    int qMax = 99;

    int force = 0;

    // Defish the image?
    float defishStrength = 0.0f;
    float defishZoom = 1.0f;

    // Input format
    enum filetype inputFiletype = FILETYPE_AUTO;

    // Whether to copy files that cannot be compressed
    int copyFiles = 1;

    // Quiet mode (less output)
    int quiet = 0;

    unsigned char *buf, *original, *originalGray = NULL, *tmpImage;
    unsigned char *compressed = NULL, *compressedGray;
    long bufSize = 0, originalSize = 0, originalGraySize = 0;
    long compressedGraySize = 0;
    unsigned long compressedSize = 0, saved;
    uint8_t *decodedImage = NULL;
    int width, height, min, max, attempt, quality;
    int jpegcs, rgb_stride, err, ok, percent, wSize;
    float metric, umetric;
    char *inputPath, *outputPath;
    FILE *file;
    WebPConfig config;
    WebPPicture pic;
    WebPMemoryWriter wrt;
    WebPMemoryWriterInit(&wrt);

    const char *optstring = "cd:fhl:m:n:q:rt:x:z:QT:V";
    static const struct option opts[] =
    {
        { "defish", required_argument, 0, 'd' },
        { "force", no_argument, 0, 'f' },
        { "help", no_argument, 0, 'h' },
        { "input-filetype", required_argument, 0, 'T' },
        { "loops", required_argument, 0, 'l' },
        { "max", required_argument, 0, 'x' },
        { "method", required_argument, 0, 'm' },
        { "min", required_argument, 0, 'n' },
        { "no-copy", no_argument, 0, 'c' },
        { "no-progressive", no_argument, 0, 'p' },
        { "ppm", no_argument, 0, 'r' },
        { "quality", required_argument, 0, 'q' },
        { "quiet", no_argument, 0, 'Q' },
        { "target", required_argument, 0, 't' },
        { "version", no_argument, 0, 'V' },
        { "zoom", required_argument, 0, 'z' },
        { 0, 0, 0, 0 }
    };
    int opt, longind = 0;

    char *progname = "webp-compress";

    while ((opt = getopt_long(argc, argv, optstring, opts, &longind)) != -1)
    {
        switch (opt)
        {
        case 'c':
            copyFiles = 0;
            break;
        case 'd':
            defishStrength = atof(optarg);
            break;
        case 'f':
            force = 1;
            break;
        case 'h':
            usage(progname);
            return 0;
        case 'l':
            attempts = atoi(optarg);
            break;
        case 'm':
            method = parseMethod(optarg);
            break;
        case 'n':
            qMin = atoi(optarg);
            break;
        case 'q':
            preset = parseQuality(optarg);
            break;
        case 'r':
            inputFiletype = FILETYPE_PPM;
            break;
        case 't':
            target = atof(optarg);
            break;
        case 'x':
            qMax = atoi(optarg);
            break;
        case 'z':
            defishZoom = atof(optarg);
            break;
        case 'T':
            if (inputFiletype != FILETYPE_AUTO)
            {
                error("multiple file types specified for the input file");
                return 1;
            }
            inputFiletype = parseInputFiletype(optarg);
            break;
        case 'Q':
            quiet = 1;
            break;
        case 'V':
            version();
            return 0;
        };
    }

    if (argc - optind != 2)
    {
        usage(progname);
        return 255;
    }

    inputPath = argv[optind];
    outputPath = argv[optind + 1];

    if (method == UNKNOWN)
    {
        error("invalid method!");
        usage(progname);
        return 255;
    }

    // No target passed, use preset!
    if (target < 0.001f)
    {
        target = setTargetFromPreset(preset);
    }

    // if (!WebPConfigPreset(&config, WEBP_PRESET_DEFAULT, 50)) {
    if (!WebPConfigPreset(&config, WEBP_PRESET_PHOTO, 50)) {
        error("could not initialize WebP configuration");
        return 1;
    }
    if (!WebPPictureInit(&pic)) {
        error("could not initialize WebP picture");
        return 1;
    }
    pic.writer = WebPMemoryWrite;
    pic.custom_ptr = (void*)&wrt;

    /* Read the input into a buffer. */
    bufSize = readFile(inputPath, (void **) &buf);
    if (!bufSize)
    {
        WebPMemoryWriterClear(&wrt);
        return 1;
    }

    /* Detect input file type. */
    if (inputFiletype == FILETYPE_AUTO)
        inputFiletype = detectFiletypeFromBuffer(buf, bufSize);

    /*
     * Read original image and decode. We need the raw buffer contents and its
     * size to obtain meta data and the original file size later.
     */
    originalSize = decodeFileFromBuffer(buf, bufSize, &original, inputFiletype, &width, &height, &jpegcs, JCS_RGB);
    if (!originalSize)
    {
        error("invalid input file: %s", inputPath);
        WebPMemoryWriterClear(&wrt);
        return 1;
    }

    if (defishStrength)
    {
        info(quiet, "Defishing...\n");
        tmpImage = malloc(width * height * 3);
        defish(original, tmpImage, width, height, 3, defishStrength, defishZoom);
        free(original);
        original = tmpImage;
    }

    // WebP image dimensions
    pic.width = width;
    pic.height = height;
    rgb_stride = width * 3;

    // Convert RGB input into Y
    originalGraySize = grayscale(original, &originalGray, width, height);

    if (!originalSize || !originalGraySize)
    {
        error("could not create the original grayscale image");
        WebPMemoryWriterClear(&wrt);
        WebPPictureFree(&pic);
    }

    if (qMin > qMax)
    {
        error("maximum JPEG quality must not be smaller than minimum quality!");
        return 1;
    }

    min = qMin;
    max = qMax;
    for (attempt = attempts - 1; attempt >= 0; --attempt)
    {
         quality = (max + min + 1) / 2;

        /* Terminate early once bisection interval is a singleton. */
        if (min == max)
            attempt = 0;

        WebPMemoryWriterClear(&wrt);

        ok = WebPPictureImportRGB(&pic, original, rgb_stride);
        if (!ok) {
            error("could not import RGB image to WebP");
            WebPMemoryWriterClear(&wrt);
            free(original);
            free(originalGray);
            return 1;
        }

        // Recompress to a new quality level
        config.quality = (float)quality;
        ok = WebPEncode(&config, &pic);
        if (!ok)
        {
            error("could not encode image to WebP");
            WebPMemoryWriterClear(&wrt);
            WebPPictureFree(&pic); // must be called independently of the 'ok' result
            free(original);
            free(originalGray);
            return 1;
        }
        compressedSize = wrt.size;

        // Decode the just encoded buffer
        decodedImage = WebPDecodeRGB(wrt.mem, wrt.size, &width, &height);
        if (decodedImage == NULL) {
            error("unable to decode buffer that was just encoded!");
            WebPMemoryWriterClear(&wrt);
            WebPPictureFree(&pic);
            free(originalGray);
            return 1;
        }

        // Convert RGB input into Y
        compressedGraySize = grayscale(decodedImage, &compressedGray, width, height);

        // Free the decoded RGB image
        WebPFree(decodedImage);

        if (!compressedGraySize)
        {
            error("unable to decode file that was just encoded!");
            WebPMemoryWriterClear(&wrt);
            WebPPictureFree(&pic);
            free(originalGray);
            return 1;
        }

        if (!attempt)
            info(quiet, "Final optimized ");

        // Measure quality difference
        metric = MetricCalc(method, originalGray, compressedGray, width, height, 1);
        umetric = MetricRescale(method, metric);
        info(quiet, MetricName(method));

        if (attempt)
            info(quiet, " at q=%i (%i - %i): UM %f\n", quality, min, max, umetric);
        else
            info(quiet, " at q=%i: UM %f\n", quality, umetric);

        if (umetric < target)
        {
            if (compressedSize >= bufSize)
            {
                free(compressed);
                free(compressedGray);

                if (copyFiles)
                {
                    info(quiet, "Output file would be larger than input!\n");
                    file = openOutput(outputPath);
                    if (file == NULL)
                    {
                        error("could not open output file: %s", outputPath);
                        return 1;
                    }

                    fwrite(buf, bufSize, 1, file);
                    fclose(file);

                    free(buf);

                    return 0;
                }
                else
                {
                    error("output file would be larger than input!");
                    free(buf);
                    return 1;
                }
            }
            min = MIN(quality, max);
        }
        else
        {
            max = MAX(quality, min);
        }

        // If we aren't done yet, then free the image data
        if (attempt)
        {
            WebPPictureFree(&pic);
        }
    }

    free(buf);

    // Calculate and show savings, if any
    percent = compressedSize * 100 / bufSize;
    saved = (bufSize > compressedSize) ? (bufSize - compressedSize) : 0;
    info(quiet, "New size is %i%% of original (saved %lu kb)\n", percent, saved / 1024);

    if (compressedSize >= bufSize && !force)
    {
        error("output file is larger than input, aborting!");
        return 1;
    }

    // Open output file for writing
    file = openOutput(outputPath);
    if (file == NULL)
    {
        error("could not open output file");
        WebPMemoryWriterClear(&wrt);
        return 1;
    }

    /* Write image data. */
    wSize = fwrite(wrt.mem, wrt.size, 1, file);

    WebPMemoryWriterClear(&wrt);

    if (wSize != 1) {
        fclose(file);
        error("could not write to output file: %s", outputPath);

        return 1;
    }

    err = fclose(file);
    if (err) {
        error("could not close the output file: %s", outputPath);
        return 1;
    }

    free(original);
    free(originalGray);

    return 0;
}
